### PR TITLE
[ROCm] fix: remove extra bf16 -> fp32 cast in jit grouped topk kernel path

### PR DESCRIPTION
**Pull request by RuibinCheung**
## Motivation

On ROCm with `SGLANG_OPT_USE_JIT_KERNEL_GROUPED_TOPK=1`, every MoE layer pays an
avoidable `bfloat16 -> float32` copy kernel on the routing bias.

`select_experts` unconditionally converts `correction_bias` into the gating dtype
whenever `_use_aiter and use_grouped_topk`:

```python
    if _use_aiter and use_grouped_topk and correction_bias is not None:
        correction_bias = topk_config.correction_bias_for_dtype(router_logits.dtype)
```

That conversion exists for `aiter_biased_grouped_topk`, which wants the bias in the
same dtype as the gating logits. But when the JIT router is enabled,
`biased_grouped_topk_gpu` returns from its `jit_grouped_gate` branch long before the
aiter branch is reached, and that branch immediately converts the bias back:

```python
        return jit_grouped_gate(
            gating_output,
            correction_bias.to(dtype=torch.float32),
            ...
        )
```

So a model that stores `e_score_correction_bias` in fp32 (GLM-5.2 does, deliberately —
its bias sits at an offset whose spread is only a few bf16 ULPs wide) gets downcast to
bf16 once, then upcast back to fp32 on **every call, for every MoE layer**. The
consumer that wanted the bf16 copy never runs.

The upcast is also redundant on its own terms: the Triton `moe_fused_gate` already
accepts a fp32/fp16/bf16 bias and `_router_triton_kernel` widens it in-register on
load, so no host-side copy is needed in either direction.

Measured on GLM-5.2-MXFP4 (`GlmMoeDsaForCausalLM`, 78 layers, 256 routed experts),
MI355X / TP8 / conc 1 / ISL 100k / MTP: **76 extra kernel launches per target-verify
forward, 0.32 ms/step.** Each launch is ~4 us of pure launch overhead for a `[256]`
tensor.

## Modifications

`python/sglang/srt/layers/moe/topk.py` — skip the downcast when the JIT grouped-topk
router is the actual consumer:

```python
    # Only aiter_biased_grouped_topk wants the bias in the gating dtype. The JIT
    # router returns before it in biased_grouped_topk_gpu and upcasts the bias
    # in-register, so downcasting here would only buy it a per-call cast back.
    if (
        _use_aiter
        and use_grouped_topk
        and correction_bias is not None
        and not envs.SGLANG_OPT_USE_JIT_KERNEL_GROUPED_TOPK.get()
    ):
        correction_bias = topk_config.correction_bias_for_dtype(router_logits.dtype)
```

Scope:

- `SGLANG_OPT_USE_JIT_KERNEL_GROUPED_TOPK` is `EnvBool(False)` by default, so the
default path is byte-for-byte unchanged — the aiter branch still receives the bias
in the gating dtype exactly as before.
- With the flag on, `correction_bias.to(dtype=torch.float32)` in the
`jit_grouped_gate` branch degenerates to a no-op for an fp32 bias, and still
converts correctly for a model that genuinely stores the bias in bf16. It is kept
rather than deleted so the branch stays correct regardless of what dtype the caller
hands it.
- The CUDA grouped-fallback branch further down `biased_grouped_topk_gpu` has the same
`correction_bias.to(...)` call and is left untouched. It was never affected by this
downcast anyway, since the downcast is guarded by `_use_aiter`.

## Accuracy Tests

The bias reaching `_router_triton_kernel` is now the original fp32 tensor instead of  
a `fp32 -> bf16 -> fp32` round trip, so if anything this removes a precision loss  
rather than introducing one. For a model whose bias is stored in bf16 the value is  
identical, since bf16 -> fp32 widening is exact.

## Speed Tests and Profiling

Setup: GLM-5.2-MXFP4, MI355X (gfx950), TP8, ROCm 7.2.26015, torch 2.9.1+rocm7.2.0,
conc 1, ISL 100k, MTP with `SGLANG_SIMULATE_ACC_LEN=3.61`,
`SGLANG_OPT_USE_JIT_KERNEL_GROUPED_TOPK=1`. Numbers are steady-state means over
decode steps 2-4 (steps 0-1 are skipped, they carry JIT warmup).

Before, per `step[TARGET_VERIFY]`:

| kernel                                           | count | time     |
| ------------------------------------------------ | ----- | -------- |
| `at::native::bfloat16tofloat32_copy_kernel_cuda` | 76    | 0.321 ms |

Expected after: 76 -> 1 per verify (only the `logits_processor` one remains), about
-0.32 ms/step.

trace:
image

## Checklist

- Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #35549170704](https://github.com/sgl-project/sglang/actions/runs/35549170704)
Latest PR Test (Extra): :x: [Run #35549170601](https://github.com/sgl-project/sglang/actions/runs/35549170601)
Latest PR Test (AMD ROCm 10): :x: [Run #35549170702](https://github.com/sgl-project/sglang/actions/runs/35549170702)

### CODE REVIEW

**Review by 1am9trash — APPROVED** [review]
LGTM.
Skip one unneeded type conversion.
Only impact amd code path.

**Comment by RuibinCheung** [comment]
stage-b-test-1-gpu-small-amd-mi35x (rocm10, linux-mi35x-gpu-1)
stage-b-test-large-8-gpu-mi35x-disaggregation-amd (rocm10, linux-mi35x-gpu-8.fabric)
These CI failed seems a known issue on main branch. It also failed on nightly CI.
image

**Comment by 1am9trash** [comment]
Seems CI is canceled frequantly. It's hard to finish a whole CI run.

image

**Comment by RuibinCheung** [comment]
Update the main branch to fix `test_disagg_idle_step_counters` and `test_auxiliary_output.py` in this PR(https://github.com/sgl-project/sglang/pull/40411).
